lib/apiconfig/load.go

```diff
@@ -81,20 +73,49 @@ func LoadClientConfigFromBytes(b []byte) (*CalicoAPIConfig, error) {
 		return nil, errors.New("invalid config file: expected kind '" + KindCalicoAPIConfig + "', got '" + c.Kind + "'")
 	}
 
+	applyConfigDefaults(&c)
 	log.Debug("Datastore type: ", c.Spec.DatastoreType)
 	return &c, nil
 }
 
-// LoadClientConfig loads the ClientConfig from the specified file (if specified)
-// or from environment variables (if the file is not specified).
+// LoadClientConfigFromEnvironment loads a client config from the environment.
+// The datastore type is defaulted if not specified.
 func LoadClientConfigFromEnvironment() (*CalicoAPIConfig, error) {
 	c := NewCalicoAPIConfig()
-
-	// Load client config from environment variables.
-	log.Debug("Loading config from environment")
 	if err := envconfig.Process("calico", &c.Spec); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to load config from env vars: %w", err)
 	}
-
+	applyConfigDefaults(c)
 	return c, nil
 }
+
+// applyConfigDefaults tries to detect the correct datastore type and config parameters.
+func applyConfigDefaults(c *CalicoAPIConfig) {
+	if c.Spec.DatastoreType == "" {
+		log.Debug("Datastore type isn't set, trying to detect it")
+		if c.Spec.EtcdEndpoints != "" {
+			log.Debug("EtcdEndpoints specified, detected etcdv3.")
+			c.Spec.DatastoreType = EtcdV3
+		} else {
+			log.Debug("No EtcdEndpoints specified, defaulting to kubernetes.")
+			c.Spec.DatastoreType = Kubernetes
+		}
+	}
+
+	if c.Spec.DatastoreType == Kubernetes {
+		// Default to using $(HOME)/.kube/config, unless another means has been configured.
+		switch {
+		case c.Spec.Kubeconfig != "":
+			log.WithField("kubeconfig", c.Spec.Kubeconfig).Debug("kubeconfig provided.")
+		case c.Spec.K8sAPIEndpoint != "":
+			log.WithField("apiEndpoint", c.Spec.K8sAPIEndpoint).Debug("API endpoint provided.")
+		case os.Getenv("HOME") == "":
+			// No home directory, can't build a default config path.
+			log.Debug("No home directory, default path doesn't apply.")
```

Comment:
Warning?  Bearing in mind that this means there won't be any effective k8s config.

Reply:
I _think_ some legitimate paths go through here (Typha maybe?) because they fill in some values after this step.